Ratpack instrumentation support

jonmort opened the pull request:
I've been working on this for quite a while and been round and round on it. This is finally ready for some feedback so please provide it.

There are several areas I'm not sure about
* Naming of the instrumentation
* Component Tag- There are two here, the main handler that implements the http response and the http client
* Resource Tag - I've left this out because I couldn't see how to build it well. I also had no idea what it should be
* Use of Java 8 Features. Ratpack only supports Java 8 so I used them
* Bytebuddy - I've not worked with this before so I tried to copy the patterns I saw. I'm particularly unsure about how the `classLoaderHasClasses()` calls are supposed to be used
* Version plugin in the gradle file - I declared all of the classes that have no compile dependency on them in the gradle file - those classes that are instrumented. This appeared to be what the other classes instrumentations did but I was not sure
* Logging - I felt like I wanted to add some more, but there isn't much elsewhere so I didn't
* Versions of Ratpack to support and test. I chose 1.4.6+ as the API is good for the integration. In Ratpack 1.6 there will be a much better way to instrument `HttpClient`, what I have works but it would be much cleaner when `HttpClient` interceptors are available
* Appropriate level of comments - I added a bit more than I found elsewhere, but they aren't extensive

Also, is there a contributors agreement I need to sign?

realark commented:
@jonmort Awesome work!

I'll get back to you about that contributors agreement and the rest of your questions early next week.

One blocker is this instrumentation is compiled against java 8. We have to support  java 1.7, and attempting to load java 8 bytecode crashes the app. We'll put out a solution for this next week so we can unblock this PR and https://github.com/DataDog/dd-trace-java/pull/254.

tylerbenson commented:
Looks like this is something we could potentially do with custom `sourceSets`: https://stackoverflow.com/a/18191391

jdbevan commented on dd-java-agent/instrumentation/ratpack/src/main/java/datadog/trace/instrumentation/ratpack/TracingHandler.java:
```diff
@@ -0,0 +1,57 @@
+              Status status = response.getStatus();
+              if (status != null) {
+                // Should a 4xx be marked as an error?
+                if (status.is4xx() || status.is5xx()) {
```
Can you expand on why that is? 4xx responses are, by definition, also errors...

jonmort commented on dd-java-agent/instrumentation/ratpack/src/test/groovy/RatpackTest.groovy:
```diff
@@ -0,0 +1,279 @@
+    span.context().tags["component"] == "handler"
+    span.context().spanType == DDSpanTypes.WEB_SERVLET
+    !span.context().getErrorFlag()
+    span.context().tags["http.url"] == "/"
```
I hadn't worked out how to do this when i first looked at it, now I see we can do this beforeSend and pull the path bindings then. I wasn't quite sure what the resource name was in ratpack, but the path binding description appears to be perfectly good

jonmort commented on dd-java-agent/instrumentation/ratpack/src/main/java/datadog/trace/instrumentation/ratpack/TracingHandler.java:
```diff
@@ -0,0 +1,57 @@
+
+    final Scope scope =
+        GlobalTracer.get()
+            .buildSpan("ratpack")
```
@tylerbenson is this span name ok?

jonmort commented:
I have addressed all the comments. I'm waiting on a solution for Java 7. I realised there was a Java 8 method reference in the instrumentation which i have replaced with a new class to ensure Java 7 byte code compatibility for the instrumentation class. The other classes have significant use of Java 8 language features.

Regarding Netty, I think it might be possible to integrate this instrumentation, but mostly when using Ratpack you don't want Netty details as the abstractions Ratpack provides are more functional

tylerbenson commented on dd-java-agent/instrumentation/ratpack-1.4/src/test/groovy/RatpackTest.groovy:
```diff
@@ -0,0 +1,299 @@
+
+    span.context().serviceName == "unnamed-java-app"
+    span.context().operationName == "ratpack"
+    span.context().resourceName == ":foo/:bar?/baz"
```
Sorry I didn't mention this before, but the resource name should ideally be prefixed with the HTTP verb, so in this case, the resource name would be `GET /:foo/:bar?/baz`. Is that possible?

jonmort replied:
Yes that makes sense. I'm also missing the leading slash. I'll put that in too

tylerbenson commented on dd-java-agent/instrumentation/ratpack-1.4/src/main/java/datadog/trace/instrumentation/ratpack/RatpackInstrumentation.java:
```diff
@@ -0,0 +1,170 @@
+                "ratpack.handling.Context",
+                "ratpack.handling.Handler",
+                "ratpack.http.Request",
+                "ratpack.http.Status"))
```
These classes should mirror what is listed in the `versionScan.verifyPresent` block.  This is how we ensure we don't try to instrument something in `1.3`

tylerbenson commented on dd-java-agent/instrumentation/ratpack-1.4/ratpack-1.4.gradle:
```diff
@@ -0,0 +1,47 @@
+versionScan {
+  group = "io.ratpack"
+  module = 'ratpack'
+  versions = "[1.4.0,)"
```
I'm ok restricting it to `1.4` as long as it doesn't try to inadvertently instrument any unsupported classes. We ensure this by including the below `verifyPresent` classes in the `classLoaderHasClasses` list in the instrumentation.

jonmort replied:
So this is where it gets complex! The reason 1.3 isn't supported is because the interfaces have different methods, they are the same classes but method signatures changed. Specifically it is the `RequestSpec` that has the changes. I could rely on a class that is only in 1.4 to force this, but one that isn't instrumented

tylerbenson replied:
Yeah, that is ok. The classes in that list aren't the ones we're explicitly instrumenting. They're intended as a "fingerprint" to identify the supported versions.

jonmort replied:
I've sorted out the fingerprint now. It is pretty much just that method on that class that identifies 1.4 from 1.3

tylerbenson commented on settings.gradle:
```diff
@@ -33,6 +33,7 @@ include ':dd-java-agent:instrumentation:osgi-classloading'
 include ':dd-java-agent:instrumentation:servlet-2'
 include ':dd-java-agent:instrumentation:servlet-3'
 include ':dd-java-agent:instrumentation:spring-web'
+include ':dd-java-agent:instrumentation:ratpack-1.4'
```
Please re-order alphabetically.

jonmort commented on dd-java-agent/instrumentation/ratpack-1.4/src/main/java/datadog/trace/instrumentation/ratpack/TracingHandler.java:
```diff
@@ -0,0 +1,65 @@
+  private static String getResourceName(Context ctx) {
+    String description = ctx.getPathBinding().getDescription();
+    if (description == null || description.isEmpty()) {
+      return ctx.getRequest().getUri();
```
This will happen when there is no handler bound to the path (I think the response will be 404 or 405).

jonmort commented on dd-java-agent/instrumentation/ratpack-1.4/ratpack-1.4.gradle:
```diff
@@ -0,0 +1,63 @@
+    "ratpack.http.client.ReceivedResponse"  : null,
+    "ratpack.http.client.RequestSpec"       : null,
+    "ratpack.http.client.StreamedResponse"  : null,
+    "ratpack.path.PathBinding"              : "getDescription",
```
This method (`getDescription`) is only available on Ratpack 1.4+. I don't instrument `PathBinding` but do the `getDescription` method on it. I couln't see a way to express that in the agent builder. Any suggestions?

tylerbenson replied:
The only classes in this list should be the ones that are unique to version 1.4+.  In some cases, just class name isn't enough to uniquely identify. That's why we also have the optional method name in there.

jonmort commented on dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/ClassLoaderMatcher.java:
```diff
@@ -30,6 +30,11 @@ private ClassLoaderMatcher() {
     return new ClassLoaderHasClassWithFieldMatcher(className, fieldName);
   }
 
+  public static ElementMatcher.Junction.AbstractBase<ClassLoader> classLoaderHasClassWithMethod(
```
@tylerbenson I went out on a limb here and added a way to check for a class that has a particular method as I could not see a way to do it otherwise. I looked at using `classLoaderHasClassWithField` but it felt too fragile to use on a Ratpack internal class.

jonmort commented on dd-java-agent/instrumentation/ratpack-1.4/src/main/java/datadog/trace/instrumentation/ratpack/RatpackInstrumentation.java:
```diff
@@ -0,0 +1,158 @@
+      new TypeDescription.Latent(
+          "ratpack.func.Action", Modifier.PUBLIC, null, Collections.emptyList());
+
+  static final ElementMatcher.Junction.AbstractBase<ClassLoader>
```
@tylerbenson this is where I use the new ClassLoader Matcher, which gets reused for all the instrumentation

realark commented:
@jonmort Thanks for your patience while we work on unblocking you. We're still working on a solution for Java 8 instrumentation, but we'll let you know as soon as we can get this merged.

jonmort commented:
I notice that `ClassLoaderMatcher` just got updated to add the same functionality that I did, but with one key difference. The method matcher that was added uses `Classes` to pick out the method arg and mine uses `String`s. I used `String`s because if we are checking for the presence of a method with a signature as a fingerprint for versions the parameter classes may not be loaded when the check is done. In short I don't think it is safe to use `Class` in this instance

tylerbenson commented:
Good point... @realark you may want to reconsider this^

realark commented:
Thanks @jonmort That's a good point. I'm adding that fix to my next play pr.

jonmort commented:
@tylerbenson I've isolated the Java 8 code now and we have green builds. Is there anything further that needs doing here?

tylerbenson commented on dd-java-agent/instrumentation/ratpack-1.4/src/main/java8/datadog/trace/instrumentation/ratpack/impl/RatpackHttpClientAdvice.java:
```diff
@@ -0,0 +1,138 @@
+
+    @Override
+    public void execute(RequestSpec requestSpec) throws Exception {
+      requestAction.execute(
```
Do we need any logic to prevent double-wrapping here?

tylerbenson commented on dd-java-agent/instrumentation/ratpack-1.4/src/test/groovy/RatpackTest.groovy:
```diff
@@ -0,0 +1,300 @@
+    /*
+    This test is somewhat convoluted and it raises some questions about how this is supposed to work
+     */
+    setup:
```
Do you think there are potential problems that still need to be resolved?

jonmort commented:
Yes ready to merge
